local: mark repository files as read-only and handle chmod errors

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
Files stored by the local backend in a repository should be marked as readonly to prevent accidental modifications. However this was accidentally removed in #1258. This PR reintroduces this behavior. In addition permissions errors for the `chmod` calls are now ignored. These could cause problems with certain mounted filesystems, see #1756 and #2157.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
closes #1756
closes #2157

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- ~~[ ] I have added documentation for the changes (in the manual)~~
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

greatroar commented:
LGTM. Looks like the Chmod originally masked off all the write bits, until it was changed to a no-op in 9842eff887f779152ec220d1eccf8e847ae0c14d.

greatroar commented on internal/backend/local/local.go:
```diff
@@ -201,15 +197,7 @@ func (b *Local) Test(ctx context.Context, h restic.Handle) (bool, error) {
-	}
-
-	return fs.Remove(fn)
+	return fs.Remove(b.Filename(h))
```
:smile:

MichaelEischer committed:
Add changelog

MichaelEischer committed:
local: Ignore permission errors on chmod call in Save/Remove operation

The file is already created with the proper permissions, thus the chmod
call is not critical. However, some file systems have don't allow
modifications of the file permissions. Similarly the chmod call in the Remove
operation should not prevent it from working.

MichaelEischer committed:
local: mark repository files as read-only

This is intended to prevent accidental modifications of data files.
Marking the files as read-only was accidentally removed in #1258.

MichaelEischer commented:
After talking with @fd0, I've turned this PR upside down. Instead of removing the previously redundant chmod call it sets the readonly flag for data files, which was accidentally remove in #1258. The permission errors from #1756 and #2157 are now ignored for the chmod calls.

rawtaz commented on changelog/unreleased/issue-1756:
```diff
@@ -0,0 +1,11 @@
+Bugfix: Mark repository files as readonly when using the local backend
+
+Files stored in a local repository are marked as readonly on non-Windows
+systems to prevent accidential file modifications. In addition, the handling
+of errors is improved which enables the local backend to work with certain
+filesystems and network mounts which do not permit modifications of file
+permissions.
```
This paragraph is unclear. The present tense makes it unclear if you are talking about how it is after this bugfix is applied, or if you are talking about the state of things before the bugfix is applied.

I suggest writing it like this instead:

~~~
Description of the problem in past tense.

Decription of the solution implemented and how it works in present tense.
~~~

Wording like "is now" can be used to convey "after the bugfix".

MichaelEischer replied:
The before and after the bugfix should be much clearer now. I've added a remark that the modifications the readonly flag should help against are something external to restic. However, I'm not really happy with the wording `modifications outside of restic.` with which I ended up with.

rawtaz commented on changelog/unreleased/issue-1756:
```diff
@@ -0,0 +1,11 @@
+Bugfix: Mark repository files as readonly when using the local backend
+
+Files stored in a local repository are marked as readonly on non-Windows
```
Might perhaps be good to write "marked as read-only in the filesystem", so it's clear where/how it's being marked as read-only.

MichaelEischer replied:
Is changed.

greatroar commented on internal/backend/local/local_unix.go:
```diff
@@ -9,6 +9,6 @@ import (
-func setNewFileMode(f string, mode os.FileMode) error {
-	return fs.Chmod(f, mode)
+func setFileReadonly(f string, mode os.FileMode) error {
+	return fs.Chmod(f, mode&os.FileMode(^uint32(0222)))
```
More modern Go would be

     mode &^ 0222

MichaelEischer replied:
I've changed it, That's indeed much more readable.

greatroar commented on internal/backend/local/local.go:
```diff
@@ -205,7 +213,7 @@ func (b *Local) Remove(ctx context.Context, h restic.Handle) error {
 
 	// reset read-only flag
 	err := fs.Chmod(fn, 0666)
```
I really don't get this one, though, at least not on Unix. Why chmod a file that we're going to remove right after?

MichaelEischer replied:
I agree that on Unix this `chmod` is pretty much useless. My idea here was to ensure that a Windows system can remove a file in a repository created on a Unix system (probably via a shared NTFS partition).

greatroar commented on changelog/unreleased/issue-1756:
```diff
@@ -0,0 +1,15 @@
+Bugfix: Mark repository files as readonly when using the local backend
+
+Files stored in a local repository were marked as writeable on the
+filesystem for non-Windows systems, which did not prevent accidential file
```
accidental

MichaelEischer replied:
I've fixed the typo

fd0 reviewed (approved):
LGTM!
